### PR TITLE
[c10d] Scalable PG initiation.

### LINKED ISSUE #81291: Lack of clear story around sub process group handling in torch.distributed
## Issue
The traditional SPMD full sync training requires all ranks to work together. This typically can work well with the regular gloo/nccl process group initialized in every rank.

Today some ML use cases require checkpointing/in-trainer publishing via a subset (1 or more in DDP/HSDP case) of ranks. A typical way is to create a sub process group using the `dist.new_group()` helper.
However, this creates challenges in various ways:

1. Currently the `new_group()` requires it to be called in all ranks (due to `_store_based_barrier()` call) to ensure the internal metadata (which are basically just some dicts) are consistent across all ranks. However, this means:
a. the sub pg cannot be created internally in some downstream libraries if those are then only invoked in those ranks. Sub pg creation (esp. gloo based) is useful for library authors to create more sophisticated logic for background processing.
b. it creates a potentially weird programming pattern where say if there are 2 calls in different sub groups, these groups' creation still needs to be hoisted like:

```
init_process_group() # world
sub_pg1 = new_group([1, 3, 5])
sub_pg2 = new_group([0])
if x:  # odd ranks
  save_checkpoint(sub_pg1)
if y:  # rank 0
  publish(sub_pg2)
```

It'd be more natural if it's like
```
init_process_group() # world
if x:  # odd ranks
  sub_pg1 = new_group([1, 3, 5])
  save_checkpoint(sub_pg1)
if y:  # rank 0
  sub_pg2 = new_group([0])
  publish(sub_pg2)
```

Note that this example may not be super obvious as it's simplified. In the real world a complicated trainer may have separate groups of people maintaining logic x and y.

2. Currently some of the calls (e.g. `broadcast()`) requires a `src` param to be passed in. This `src` somehow needs to be the global rank rather than the rank within the current pg. So in the case of odd ranks used, passing 0 would simply just stall the whole thing. Instead, the first rank within that group needs to be passed in which is `1`. However, there's no public methods for that except the internal `_get_global_rank()`. Even if that's possible it's weird to use. A group local rank may be more suitable here semantically.

## Proposal

The general proposal is to clarify the story around sub pg handling:
1. If some logic is needed in a subset of ranks, should caller pass a sub pg and invoke the call in that subset only (as opposed to invoke it in every rank and let non participating ranks return immediately)?
2. If a rank number is needed for certain collective calls, should the number be local to the process group?
3. Is a sub pg eligible to be created and destroyed in a subset of ranks (and updating metadata only in that sub group is allowed - this is currently not legit in existing API)?

This proposal would like to answer yes to all of the above.

In order to keep backward compatibility, the `new_group()` can be kept as is,
2 new APIs:

`new_sub_group()`: similar to `new_group()` but can be invoked in a subset of ranks.
`clone(group)`: creates a sub process group with the same parameters as the input unless specified.

can be introduced to explicitly allow for such capability.

https://gist.github.com/xunnanxu/1cc1ea37a81d10f15339f7aca80a38bb is a hacky version of allowing sub pg to be created within a library (`SPMDCounter` in this case) to demonstrate the technical possibility.

In that gist:
- `test_do_stuff_with_sub_pg` is the entry point creating 4 processes.
- The body `_work()` uses the public api `new_group()` to create the 1st sub group for odd ranks.
- Then the library `SPMDCounter` creates an internal pg, using internal methods, in the odd ranks only, for illustrative purpose for internal sync.

### Alternatives

_No response_

### Additional context

_No response_

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

### PR DESCRIPTION
**Pull request by kumpera**
Add use_local_synchronization argument to new_group.

When this argument is True, is change new_group to do a store_barrier only on the ranks that are park of the group and not the whole cluster.

This addressess both scalability and composability problems associated with new_group.

Fixes #81291.

This is relanding #84224
As part of the original PR I did a quick benchmark of creating 3 PGs per rank using both functions and perf is the following:

new_group use_local_synchronization=False:
| World Size | Time (in secs) |
| --- | ----------- |
| 4 | 0.12 |
| 8 | 0.25 |
| 16 | 0.51 |
| 32 | 0.87 |
| 64 | 1.50 |
| 128 | 2.87 |

new_group use_local_synchronization=True:
| World Size | Time (in secs) |
| --- | ----------- |
| 4 | 0.05 |
| 8 | 0.04 |
| 16 | 0.03 |
| 32 | 0.03 |
| 64 | 0.04 |
| 128 | 0.04 |

Scaling for `use_local_synchronization=False` is sub linear because the number of process groups created as a multiple of world_size decreases as we go up. It's 6 with world_size 4 and 192 with world_size 128.

Scaling for `use_local_synchronization=True` is constant as the number of store barriers executed per rank remains constant at 3.

Setup:

1 AWS host, backend gloo.

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -3479,7 +3483,28 @@ def _create_process_group_wrapper(
...
     return wrapped_pg
 
-def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=None):
+
+def _process_group_name(ranks, use_hashed_name):
+    global _world
+    if use_hashed_name:
+        pg_name = hashlib.sha1(bytes("_".join(map(str, ranks)), "utf-8")).hexdigest()
+        while pg_name in _world.pg_names:
+            pg_name = hashlib.sha1(bytes(pg_name + "_", "utf-8")).hexdigest()
```
Comment (xw285cornell) [inline torch/distributed/distributed_c10d.py:3492]: qq, would it be a good idea to give a readable name instead of a sha1?

**Comment by xw285cornell** [comment]
Looks great and CI all green! I'm importing it to run some tests. Feel free to commandeer it

**Review by xw285cornell — APPROVED** [review]

**Comment by dannyel2511** [comment]
Hey @kumpera, this new feature seems very interesting, I have a few questions:
1. Does this means that subprocess groups can be potentially created in parallel? For example, if we want to create these 4 groups: `[0, 1] [2, 3] [4, 5] [6, 7]` are they going to be created at the same time? I think so because now the barrier is going to be called only on participant ranks, so each group does not depend on any other. 
2. If 1) is true and subprocess groups are created in parallel, is there any risk associated to it? Or any constrain that needs to be followed?
3. When is this feature going to be released?

Thank you!

**Pull request closed without merging**